### PR TITLE
(vsphere-iso) add ability to define multiple disks

### LINKED ISSUE #8749: vsphere-iso multiple disk support
After adding the possibility to add multiple nics (which was requested in #8739) to a VM in vsphere-iso it would be very useful to get same functionality concerning disks.

For this feature there is still a pull request in the old project from JetBrains: https://github.com/jetbrains-infra/packer-builder-vsphere/pull/113/files

#### Feature Description

Adding multiple disks to a VM in the config file

#### Use Case(s)

Create a separate OS and SWAP Disks for VMs and configure them during the os installation process.

### PR DESCRIPTION
**Pull request by jhawk28**
Left the old fields so that this is a backwards compatible change. Will create a separate pull request to remove deprecated fields.

Closes #8749

### CODE REVIEW

**Comment by jhawk28** [comment]
@sylviamoss how do I kick off the build again? it looks like the one just took too long.

File: builder/vsphere/iso/step_create.go [hunk 1]
```diff
@@ -44,6 +53,8 @@ type CreateConfig struct {
 	DiskThinProvisioned bool `mapstructure:"disk_thin_provisioned"`
 	// Enable VMDK eager scrubbing for VM. Defaults to `false`.
 	DiskEagerlyScrub bool `mapstructure:"disk_eagerly_scrub"`
+	// Storage
```
[the remark this answers is not in the document]
Comment (jhawk28) [reply builder/vsphere/iso/step_create.go]: added this...

File: website/source/docs/builders/vsphere-iso.html.md.erb [hunk 2]
```diff
@@ -70,6 +70,12 @@ necessary for this build to succeed and can be found further down the page.
 ### Network Adapter Configuration
 <%= partial "partials/builder/vsphere/iso/NIC-required" %>
 
+### Storage Configuration
+<%= partial "partials/builder/vsphere/iso/DiskConfig-required" %>
+
+### Optional Storage Configuration
+<%= partial "partials/builder/vsphere/iso/DiskConfig-not-required" %>
+
```
[the remark this answers is not in the document]
Comment (jhawk28) [reply website/source/docs/builders/vsphere-iso.html.md.erb]: added this

File: builder/vsphere/driver/vm.go [hunk 3]
```diff
@@ -69,6 +66,13 @@ type CreateConfig struct {
 	USBController bool
 	Version       uint   // example: 10
 	Firmware      string // efi or bios
+	Storage       []Disk
```
[the remark this answers is not in the document]
Comment (jhawk28) [reply builder/vsphere/driver/vm.go]: I was planning on adding the acceptance test as part of the removal of the old fields. I'm not sure the acceptance tests are even running at the moment.

**Comment by jhawk28** [comment]
@azr @sylviamoss @SwampDragons Should I merge or rebase off master to get the latest? It looks like the check-generate is failing.

**Comment by SwampDragons** [comment]
I normally rebase.

File: builder/proxmox/config.go [hunk 4]
```diff
@@ -213,6 +214,9 @@ func (c *Config) Prepare(raws ...interface{}) ([]string, error) {
 	if c.Node == "" {
 		errs = packer.MultiErrorAppend(errs, errors.New("node must be specified"))
 	}
+	if strings.ContainsAny(c.TemplateName, " ") {
+		errs = packer.MultiErrorAppend(errs, errors.New("template_name must not contain spaces"))
+	}
```
Comment (carlpett) [inline builder/proxmox/config.go]: Has there been a merge the wrong direction or something in the PR? This change was merged into master in #8799.
Also see that there's a bunch of other changes which don't seem related to the vsphere-iso builder.

Comment (jhawk28) [reply builder/proxmox/config.go]: I think I rebased incorrectly. Trying again...I think it looks better now.

File: builder/vsphere/driver/vm.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jhawk28) [reply builder/vsphere/driver/vm.go]: I created issue #8841 to get the acceptance test conversation going.

[the remark this answers is not in the document]
Comment (nywilken) [reply builder/vsphere/driver/vm.go]: @jhawk28 thanks for making the changes and creating the acceptance test issue; apologies for the delayed response here. 

In speaking with the team we are going to let this PR sit for a bit until we can get the acceptance tests running. The changes look good as they are but we want to make sure we are properly setup to test new additions to the code base before merging in enhancements or major fixes. 

I know this might delay the additional changes you have in mind, but it helps us in ensuring things are working as good, if not better, than they are now. Thanks again for pushing through this change and for sticking with us as we get the acceptance testing squared away.

**Comment by SwampDragons** [comment]
I was able to stand up a vsphere cluster to test today. This looks good and works as expected. Thanks for this contribution, and thanks for your patience!!

**Comment by Jetersen** [comment]
How is this intended to be written in HCL?

`storage = []` is unsupported

```
packer -v
1.5.5
```

**Comment by Jetersen** [comment]
Seems like a bug that `disk_size` is still required

So this how it would look
```hcl
  disk_size = 32768
  disk_thin_provisioned = true

  storage {
    disk_size = 16384
    disk_thin_provisioned = true
  }
  storage {
    disk_size = 8192
    disk_thin_provisioned = true
  }
```

**Comment by jhawk28** [comment]
I'm assuming that your example parses without the disk_size. There is a disk_size check in prepare that should be updated:

```go
	if c.DiskSize == 0 {
		errs = append(errs, fmt.Errorf("'disk_size' is required"))
	}
```

**Comment by Jetersen** [comment]
the example parses but if I wanted to use only storage blocks than it currently complains yes.

So this would fail because of the above code you pasted.
```
  storage {
    disk_size = 32768
    disk_thin_provisioned = true
  }
  storage {
    disk_size = 16384
    disk_thin_provisioned = true
  }
  storage {
    disk_size = 8192
    disk_thin_provisioned = true
  }
```

**Comment by Jetersen** [comment]
The check should be updated to check if storage length is zero and DiskSize is zero :)

**Comment by jhawk28** [comment]
maybe like this: https://github.com/hashicorp/packer/pull/8975
